From 2847df72871f5202974e004cec9c6535e2c76250 Mon Sep 17 00:00:00 2001
From: Xiang Chen <chenxiang66@hisilicon.com>
Date: Wed, 23 Apr 2025 17:16:26 +0800
Subject: [PATCH 566/960] downstream: kvm: hisi_virt: Register ipiv exception
 interrupt

When one of the following conditions occurs:
1. The index of VM table exceeds the supported range.
2. Guest sends SGI with IRM=1.
3. Guest sends multicast.
it triggers a exception interrupt (PPI interrupt). Just printk
exception info in interrupt handler.

Signed-off-by: Xiang Chen <chenxiang66@hisilicon.com>
Signed-off-by: Jinqian Yang <yangjinqian1@huawei.com>
Signed-off-by: Slim6882 <15605196882@163.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/kvm/vgic/vgic-init.c | 36 +++++++++++++++++++++++++++++++++
 1 file changed, 36 insertions(+)

diff --git a/arch/arm64/kvm/vgic/vgic-init.c b/arch/arm64/kvm/vgic/vgic-init.c
index f0633b9eb8fa..7004e1fc3ce7 100644
--- a/arch/arm64/kvm/vgic/vgic-init.c
+++ b/arch/arm64/kvm/vgic/vgic-init.c
@@ -3,6 +3,7 @@
  * Copyright (C) 2015, 2016 ARM Ltd.
  */
 
+#include <linux/acpi.h>
 #include <linux/uaccess.h>
 #include <linux/interrupt.h>
 #include <linux/cpu.h>
@@ -511,17 +512,30 @@ int kvm_vgic_map_resources(struct kvm *kvm)
 	return ret;
 }
 
+extern struct static_key_false ipiv_enable;
+static int ipiv_irq;
+
 /* GENERIC PROBE */
 
 void kvm_vgic_cpu_up(void)
 {
 	enable_percpu_irq(kvm_vgic_global_state.maint_irq, 0);
+	if (static_branch_unlikely(&ipiv_enable))
+		enable_percpu_irq(ipiv_irq, 0);
 }
 
 
 void kvm_vgic_cpu_down(void)
 {
 	disable_percpu_irq(kvm_vgic_global_state.maint_irq);
+	if (static_branch_unlikely(&ipiv_enable))
+		disable_percpu_irq(ipiv_irq);
+}
+
+static irqreturn_t vgic_ipiv_irq_handler(int irq, void *data)
+{
+	kvm_info("IPIV irq handler!\n");
+	return IRQ_HANDLED;
 }
 
 static irqreturn_t vgic_maintenance_handler(int irq, void *data)
@@ -631,5 +645,27 @@ int kvm_vgic_hyp_init(void)
 	}
 
 	kvm_info("vgic interrupt IRQ%d\n", kvm_vgic_global_state.maint_irq);
+
+	if (static_branch_unlikely(&ipiv_enable)) {
+		ipiv_irq = acpi_register_gsi(NULL, 18, ACPI_EDGE_SENSITIVE,
+			ACPI_ACTIVE_HIGH);
+		if (ipiv_irq < 0) {
+			kvm_err("No ipiv exception irq\n");
+			free_percpu_irq(kvm_vgic_global_state.maint_irq,
+					kvm_get_running_vcpus());
+			return -ENXIO;
+		}
+
+		ret = request_percpu_irq(ipiv_irq, vgic_ipiv_irq_handler,
+				 "ipiv exception", kvm_get_running_vcpus());
+		if (ret) {
+			kvm_err("Cannot register interrupt %d\n", ipiv_irq);
+			free_percpu_irq(kvm_vgic_global_state.maint_irq,
+					kvm_get_running_vcpus());
+			acpi_unregister_gsi(18);
+			return ret;
+		}
+	}
+
 	return 0;
 }
-- 
2.43.0

